### PR TITLE
[mypyc] feat: specialize isinstance for tuple of primitive types

### PR DESCRIPTION
**Pull request by BobTheBuidler**
This PR specializes isinstance calls where the type argument is a tuple of primitive types.

We can skip tuple creation and the associated refcounting, and daisy-chain the primitive checks with an early exit option at each step.

### CODE REVIEW

**Comment by BobTheBuidler** [comment]
Not really sure where a run test best fits. Maybe the IR test is straightforward enough?

File: mypyc/irbuild/specialize.py [hunk 1]
```diff
@@ -587,26 +587,79 @@ def translate_isinstance(builder: IRBuilder, expr: CallExpr, callee: RefExpr) ->
...
 
-    if isinstance(expr.args[1], (RefExpr, TupleExpr)):
-        builder.types[expr.args[0]] = AnyType(TypeOfAny.from_error)
+    obj_expr = expr.args[0]
+    type_expr = expr.args[1]
 
-        irs = builder.flatten_classes(expr.args[1])
+    if isinstance(type_expr, TupleExpr) and not type_expr.items:
+        # we can compile this case to a noop
+        return builder.false()
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:596]: Is there a test case for this?

File: mypyc/test-data/irbuild-isinstance.test [hunk 2]
```diff
@@ -189,3 +189,31 @@ def is_tuple(x):
 L0:
     r0 = PyTuple_Check(x)
     return r0
+
+[case testTupleOfPrimitives]
+from typing import Any
+
+def is_instance(x: Any) -> bool:
+    return isinstance(x, (str, int, bytes))
```
Comment (JukkaL) [inline mypyc/test-data/irbuild-isinstance.test:197]: Test also single-item tuple as an edge case.

**Review by JukkaL — COMMENTED** [review]
What about adding a run test to `run-misc.test`?

**Comment by BobTheBuidler** [comment]
Both of those just as run tests, or IR too?

**Comment by BobTheBuidler** [comment]
Added run tests for various cases. no additional IR tests at this time.

File: mypyc/irbuild/specialize.py [hunk 3]
```diff
@@ -587,26 +587,79 @@ def translate_isinstance(builder: IRBuilder, expr: CallExpr, callee: RefExpr) ->
...
+        pass_block = BasicBlock()
+        fail_block = BasicBlock()
+        exit_block = BasicBlock()
+
+        # Chain the checks: if any succeed, jump to pass_block; else, continue
+        for i, desc in enumerate(descs):
+            is_last = i == len(descs) - 1
+            next_block = fail_block if is_last else BasicBlock()
+            builder.add_bool_branch(
+                builder.primitive_op(desc, [obj], expr.line), pass_block, next_block  # type: ignore [arg-type]
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py]: Can you avoid the `# type: ignore`? If there is some typing issue, I'd prefer a `cast` since it's more specific and tells which argument is the problematic one.

**Review by JukkaL — COMMENTED** [review]
Thanks for adding the tests! Looks mostly good now, just one thing remains.

File: mypyc/irbuild/specialize.py (hunk 3 above)
Comment (BobTheBuidler) [reply mypyc/irbuild/specialize.py]: Done. It was because `descs` has type `list[PrimitiveDescription | None]`

I'm assuming the `if None in descs: return None` can probably be considered as a TypeGuard of sorts by mypy but that's above my paygrade

File: mypyc/irbuild/specialize.py [hunk 4]
```diff
@@ -587,26 +588,81 @@ def translate_isinstance(builder: IRBuilder, expr: CallExpr, callee: RefExpr) ->
...
+        pass_block = BasicBlock()
+        fail_block = BasicBlock()
+        exit_block = BasicBlock()
+
+        # Chain the checks: if any succeed, jump to pass_block; else, continue
+        for i, desc in enumerate(descs):
+            is_last = i == len(descs) - 1
+            next_block = fail_block if is_last else BasicBlock()
+            builder.add_bool_branch(
+                builder.primitive_op(cast(PrimitiveDescription, desc), [obj], expr.line),
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:645]: Here a slightly better way to narrow the type would be to use `assert desc is not None` before the call, but this is fine as well.
